client/components/admin_dashboard/Grid.jsx

```diff
@@ -52,81 +19,21 @@ const Toolbar = () => (
   </GridToolbarContainer>
 );
 
-export default function Grid({ rows }) {
+export default function Grid({ rows, refreshData }) {
   const [searchText, setSearchText] = useState("");
-  const [customRows, setCustomRows] = useState(rows);
 
-  const columns = [
-    {
-      field: "visitOrder",
-      headerName: "Visit Order",
-      width: 190,
-      valueGetter: (_, row) => row.schedule.order,
-    },
-    {
-      field: "markVisited",
-      headerName: "Visited?",
-      width: 190,
-      renderCell: (params) => {
-        const { id, status } = params.row;
-
-        return (
-          <Button
-            variant={status === "Visited" ? "contained" : "outlined"}
-            color="primary"
-            onClick={() => toggleVisited(id)}
-          >
-            {status === "Visited" ? "Visited" : "Not Visited"}
-          </Button>
-        );
-      },
-    },
-    {
-      field: "status",
-      headerName: "Status",
-      width: 190,
-      renderCell: (status) => <StatusChange />,
-    },
-    {
-      valueFormatter: formatName,
-      field: "name",
-      headerName: "Name",
-      width: 190,
-    },
-    {
-      valueFormatter: formatDateCreated,
-      field: "dateCreated",
-      headerName: "Requested on",
-      width: 190,
-    },
-    {
-      valueFormatter: formatTimeRange,
-      field: "timeRange",
-      headerName: "Timeslot",
-      width: 190,
-    },
-    {
-      valueFormatter: formatPhone,
-      field: "phone",
-      headerName: "Phone",
-      width: 190,
-    },
-    { field: "email", headerName: "Email", width: 190 },
-    {
-      field: "address",
-      headerName: "Address",
-      width: 190,
-    },
-  ];
-
-  const toggleVisited = async (id) => {
-    await updateVisitedOnServer(id);
-    const updatedRows = await fetchAppointments();
-    setCustomRows(updatedRows);
+  const toggleVisited = async (id, status) => {
+    await updateStatusOnServer(id, status);
+    await refreshData();
   };
 
+  const columns = useMemo(
+    () => getColumns(refreshData, toggleVisited),
```

Comment:
```suggestion
    () => getColumns(refreshData),
```
This structure is quite weird. Not sure why we define toggleVisited here instead of inside columns.jsx?